### PR TITLE
User story 16

### LINKED ISSUE #20: User Story 16, User Show Page
```
As a Visitor,
When I click on a user's name for any book review
I am taken to a show page for that user.
I should see all reviews that this
user has written.
Each review shows:
- the title of the review
- the description of the review
- the rating of the review
- the title of the book
- the thumbnail image for the book
- the date the review was written
```

DO THESE STORIES FIRST:
- User Story 13, Add a Review for a Book

### Checklist:

- [x] base code tests are written
- [x] base code is written to pass tests
- [x] edge case ("sad path") tests written
- [x] edge case code written
- [x] code is reviewed
- [x] branch is merged
- [x] all tests still pass

child of #12

child of #33

depends on #16

### PR DESCRIPTION
**Pull request by DavisC0801**
This PR\Branch...
  Resolves #20
  adds a show page view for Users, which list all reviews that a user has written.
  adds a user controller
  adds a route for user show page

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class UsersController < ApplicationController
+  def show
+    if User.pluck(:id).include?(params[:id].to_i)
+      @user = User.find(params[:id])
+    else
+      redirect_to "/books"
```
Comment (chakeresa) [inline app/controllers/users_controller.rb]:
```suggestion
      redirect_to root
```

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<div id="review-list">
+  <% @user.reviews.each do |review| %>
+  <div id="book-<%= review.book.id %>-info">
+    <img src="<%= review.book.thumbnail %>" alt="Cover of <%= review.book.title %>">
+    <br><h3><%= review.book.title %></h3><br>
+    <!-- possible format change for readability -->
+  </div>
+  <div id="review-<%= review.id %>-info">
+    <h3><%= review.title %></h3>
+    <p>Written by: <%= @user.username %></p>
```
Comment (chakeresa) [inline app/views/users/show.html.erb]:
```suggestion

```

Comment (chakeresa) [reply app/views/users/show.html.erb]: Since we're already on a particular user's show page, I don't think we need to include this for each review.

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<div id="review-list">
+  <% @user.reviews.each do |review| %>
+  <div id="book-<%= review.book.id %>-info">
```
Comment (chakeresa) [inline app/views/users/show.html.erb]: Let's indent all the contents of this each block for readability

File: spec/features/users/show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        expect(page).to have_content(@review_2.title)
+        expect(page).to have_content("Written by: #{@user_1.username}")
+        expect(page).to have_content("Written on: #{Date.strptime(@review_2.updated_at.to_s)}")
+        expect(page).to have_content("Rating: #{@review_2.rating}/5")
+        expect(page).to have_content(@review_2.text)
+      end
+    end
+
+    describe "when I give an invalid user_id" do
+      it "redirects me to the book index page" do
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb:69]:
```suggestion
      it "redirects me to the root page" do
```

File: spec/features/users/show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        visit user_path(@user_1.id + 50)
+
+        expect(current_path).to eq(books_path)
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb:72]:
```suggestion
        expect(current_path).to eq(root)
```

File: app/views/users/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  <div id="review-<%= review.id %>-info">
+    <h3><%= review.title %></h3>
+    <p>Written by: <%= @user.username %></p>
+    <p>Written on: <%= Date.strptime(review.updated_at.to_s)%></p>
+    <p>Rating: <%= review.rating %>/5</p>
+    <p><%= review.text %></p>
+  </div>
+  <% end %>
+  <% if @user.reviews.count == 0 %>
+  <p>This user has written no reviews yet.</p>
```
Comment (chakeresa) [inline app/views/users/show.html.erb]: Indent contents of this if block for readability

File: spec/features/users/show_spec.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
+require "rails_helper"
+
+RSpec.describe "as a visitor" do
+  describe "when I click on a user's name for any book review" do
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb:4]: hm, this doesn't seem relevant for all the tests inside it, only for the first one -- consider moving that one out & adding a different describe block for the majority of the show page tests

File: spec/features/users/show_spec.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      @user_1 = User.create(username: "Chris Davis")
+      @user_2 = User.create(username: "Alexandra Chakeres")
+      @review_1 = @book_1.reviews.create!(title: "This book rocks!", rating: 5, text: "Read it!", user: @user_1)
+      @review_2 = @book_2.reviews.create!(title: "This book sucks!", rating: 1, text: "Don't read it!", user: @user_1)
+    end
+
+    it "i am taken to a show page for that user." do
+      visit book_path(@book_1)
+
+      click_link("#{@user_1.username}")
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb]:
```suggestion
      click_link(@user_1.username)
```

File: spec/features/users/show_spec.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      visit(user_path(@user_1))
+
+      within("#book-#{@book_1.id}-info") do
+        expect(page).to have_css("img[src*='#{@book_1.thumbnail}']")
+        expect(page).to have_content(@book_1.title)
+      end
+
+      within("#review-#{@review_1.id}-info") do
+        expect(page).to have_content(@review_1.title)
+        expect(page).to have_content("Written by: #{@user_1.username}")
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb]:
```suggestion
```

File: spec/features/users/show_spec.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        expect(current_path).to eq(books_path)
+      end
+    end
+
+    describe "when I visit a user page that has no reviews" do
+      it "has no content" do
+        visit user_path(@user_2.id)
+        
+        within("#review-list") do
+          expect(page).to have_content("This user has written no reviews yet.")
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb]:
```suggestion
          expect(page).to have_content("This user has not written any reviews yet.")
```

File: app/views/users/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  </div>
+  <div id="review-<%= review.id %>-info">
+    <h3><%= review.title %></h3>
+    <p>Written by: <%= @user.username %></p>
+    <p>Written on: <%= Date.strptime(review.updated_at.to_s)%></p>
+    <p>Rating: <%= review.rating %>/5</p>
+    <p><%= review.text %></p>
+  </div>
+  <% end %>
+  <% if @user.reviews.count == 0 %>
```
Comment (chakeresa) [inline app/views/users/show.html.erb]:
```suggestion
  <% if @user.review_count == 0 %>
```

Comment (chakeresa) [reply app/views/users/show.html.erb]: Calling `reviews.count` violates MVC because `.count` is a class method :-(. So we'll have to add a class method `review_count` to our `User` class (which will just be `reviews.count` & the associated test(s).

File: app/views/users/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,19 @@
+<div id="review-list">
```
Comment (chakeresa) [inline app/views/users/show.html.erb:1]:
```suggestion
<h1><%= @user.username %></h1>
<div id="review-list">
```

File: spec/features/users/show_spec.rb [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+      click_link("#{@user_1.username}")
+
+      expect(page.status_code).to eq(200)
+      expect(current_path).to eq(user_path(@user_1))
+    end
+
+    it "the user page shows all information for all reviews the user has written" do
+      visit(user_path(@user_1))
+
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb:37]:
```suggestion

expect(page).to have_content(@user.username)
```

File: spec/features/users/show_spec.rb [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      end
+
+      within("#book-#{@book_2.id}-info") do
+        expect(page).to have_css("img[src*='#{@book_2.thumbnail}']")
+        expect(page).to have_content(@book_2.title)
+      end
+
+      within("#review-#{@review_2.id}-info") do
+        expect(page).to have_content(@review_2.title)
+        expect(page).to have_content("Written by: #{@user_1.username}")
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb]:
```suggestion
```

File: spec/features/users/show_spec.rb [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    describe "when I give an invalid user_id" do
+      it "redirects me to the book index page" do
+        visit user_path(@user_1.id + 50)
+
+        expect(current_path).to eq(books_path)
+      end
+    end
+
+    describe "when I visit a user page that has no reviews" do
+      it "has no content" do
```
Comment (chakeresa) [inline spec/features/users/show_spec.rb]:
```suggestion
      it "alerts the user that there are no reviews" do
```

**Comment by chakeresa** [comment]
@DavisC0801 I made some edits / comments on the code for this PR in GitHub. Let's review & resolve them together and then we can do the merge.

File: spec/features/users/show_spec.rb (hunk 4 above)
Comment (DavisC0801) [reply spec/features/users/show_spec.rb:69]: Right now the root page is a redirect to book index, and we can change the target in the view just as easily.

File: spec/features/users/show_spec.rb (hunk 7 above)
Comment (DavisC0801) [reply spec/features/users/show_spec.rb:4]: Agreed, fixed in branch

File: app/views/users/show.html.erb (hunk 6 above)
Comment (DavisC0801) [reply app/views/users/show.html.erb]: Agreed, formatting changes applied

File: spec/features/users/show_spec.rb (hunk 8 above)
Comment (DavisC0801) [reply spec/features/users/show_spec.rb]: String is needed for click_link, changed to a to_s call.

File: spec/features/users/show_spec.rb (hunk 9 above)
Comment (DavisC0801) [reply spec/features/users/show_spec.rb]: Agreed, changed on branch.

File: app/views/users/show.html.erb (hunk 11 above)
Comment (DavisC0801) [reply app/views/users/show.html.erb]: Noted, fixed on branch
